helix-core/src/main/java/org/apache/helix/controller/stages/MessageDispatchStage.java

```diff
@@ -78,7 +78,13 @@ protected void processEvent(ClusterEvent event, MessageOutput messageOutput) thr
         batchMessage(dataAccessor.keyBuilder(), messagesToSend, resourceMap, liveInstanceMap,
             manager.getProperties());
 
-    List<Message> messagesSent = sendMessages(dataAccessor, outputMessages);
+    String expectedSession = event.getAttribute(AttributeName.CONTROLLER_LEADER_SESSION.name());
+    // An early check for expected leader session. If the sessions don't match, it means the
+    // controller lost leadership, then messages should not be sent. This potentially avoid
+    // double masters for a single partition.
+    List<Message> messagesSent = manager.getSessionId().equals(expectedSession)
```

Comment:
Shall we throw an exception to break the pipeline? If we know session is not matched. Then there is no reason to keep following pipeline based on stale data.

Reply:
I proposed to do some checks at stages like `ReadClusterDataStage` and terminate the remaining pipeline if controller loses leadership. In our issue, I see that leader lost leadership at `ReadClusterDataStage`. So there is no need to continue the pipeline and it would help save some computation resource or write traffic to EV/persist assignment. But in the discussion we agreed on not doing that and just blocking messages. We could definitely pick it up and discuss again. @jiajunwang

Reply:
I've updated it to stop the pipeline at this point: thrown StageException.